File: utils/cronTracker.js
/**
 * Utility module for tracking cron job execution in Supabase.
 */

/**
 * Start tracking a cron job run.
 * 
 * @param {Object} supabaseClient - Initialized Supabase client
 * @param {string} jobName - Name of the job (e.g., "us_caselaw_scraper")
 * @returns {Promise<Object>} - Object containing { runId, success }
 */
export async function startJob(supabaseClient, jobName) {
    try {
        const { data, error } = await supabaseClient
            .from('cron_job_runs')
            .insert([{
                job_name: jobName,
                start_time: new Date().toISOString(),
                status: 'started',
                new_cases_found: 0,
                pages_processed: 0
            }])
            .select('id')
            .single();

        if (error) {
            console.error(`Error starting cron job tracking: ${error.message}`);
            return { runId: null, success: false };
        }

        console.log(`Cron job '${jobName}' started. Run ID: ${data.id}`);
        return { runId: data.id, success: true };
    } catch (e) {
        console.error(`Exception in cron job tracking: ${e.message}`);
        return { runId: null, success: false };
    }
}

/**
 * Mark a job as completed with metrics.
 * 
 * @param {Object} supabaseClient - Initialized Supabase client
 * @param {string} runId - ID of the job run from startJob
 * @param {Object} metrics - Dictionary of metrics (e.g., {newCasesFound: 10, pagesProcessed: 5})
 * @returns {Promise<boolean>} - Success status
 */
export async function completeJob(supabaseClient, runId, metrics) {
    if (!runId) {
        console.warn('Warning: Cannot complete job, no runId provided.');
        return false;
    }

    try {
        const updateData = {
            end_time: new Date().toISOString(),
            status: 'completed',
            ...metrics
        };

        const { error } = await supabaseClient
            .from('cron_job_runs')
            .update(updateData)
            .eq('id', runId);

        if (error) {
            console.error(`Error completing cron job: ${error.message}`);
            return false;
        }

        return true;
    } catch (e) {
        console.error(`Exception completing cron job: ${e.message}`);
        return false;
    }
}

/**
 * Mark a job as failed with error details.
 * 
 * @param {Object} supabaseClient - Initialized Supabase client
 * @param {string} runId - ID of the job run from startJob
 * @param {Object} metrics - Dictionary of metrics (e.g., {newCasesFound: 10, pagesProcessed: 5})
 * @param {string} errorMessage - Description of the error that caused the failure
 * @returns {Promise<boolean>} - Success status
 */
export async function failJob(supabaseClient, runId, metrics, errorMessage) {
    if (!runId) {
        console.warn('Warning: Cannot fail job, no runId provided.');
        return false;
    }

    try {
        const updateData = {
            end_time: new Date().toISOString(),
            status: 'failed',
            error_message: errorMessage,
            ...metrics
        };

        const { error } = await supabaseClient
            .from('cron_job_runs')
            .update(updateData)
            .eq('id', runId);

        if (error) {
            console.error(`Error marking cron job as failed: ${error.message}`);
            return false;
        }

        return true;
    } catch (e) {
        console.error(`Exception marking cron job as failed: ${e.message}`);
        return false;
    }
} 

File: utils/cron_tracker.py
"""
Utility module for tracking cron job execution in Supabase.
"""

import os
import time
from typing import Any, Dict, Optional, Tuple, Union


def start_job(supabase_client, job_name: str) -> Tuple[Optional[str], bool]:
    """
    Start tracking a cron job run.

    Args:
        supabase_client: Initialized Supabase client
        job_name: Name of the job (e.g., "singapore_caselaw_scraper")

    Returns:
        Tuple of (job_run_id, success)
    """
    try:
        start_time = time.strftime("%Y-%m-%d %H:%M:%S%z")
        response = (
            supabase_client.table("cron_job_runs")
            .insert(
                {
                    "job_name": job_name,
                    "start_time": start_time,
                    "status": "started",
                    "new_cases_found": 0,
                    "pages_processed": 0,
                }
            )
            .execute()
        )

        if response.data:
            job_run_id = response.data[0]["id"]
            print(f"Cron job '{job_name}' started. Run ID: {job_run_id}")
            return job_run_id, True
        else:
            print(f"Error: Could not create cron_job_runs record for {job_name}.")
            return None, False
    except Exception as e:
        print(f"Error starting cron job tracking: {e}")
        return None, False


def complete_job(supabase_client, job_run_id: str, metrics: Dict[str, Any]) -> bool:
    """
    Mark a job as completed with metrics.

    Args:
        supabase_client: Initialized Supabase client
        job_run_id: ID of the job run from start_job
        metrics: Dictionary of metrics (e.g., {"new_cases_found": 10, "pages_processed": 5})

    Returns:
        Success status (bool)
    """
    if not job_run_id:
        print("Warning: Cannot complete job, no job_run_id provided.")
        return False

    try:
        update_data = {
            "end_time": time.strftime("%Y-%m-%d %H:%M:%S%z"),
            "status": "completed",
            **metrics,
        }

        response = (
            supabase_client.table("cron_job_runs")
            .update(update_data)
            .eq("id", job_run_id)
            .execute()
        )

        return bool(response.data)
    except Exception as e:
        print(f"Error completing cron job: {e}")
        return False


def fail_job(
    supabase_client, job_run_id: str, metrics: Dict[str, Any], error_message: str
) -> bool:
    """
    Mark a job as failed with error details.

    Args:
        supabase_client: Initialized Supabase client
        job_run_id: ID of the job run from start_job
        metrics: Dictionary of metrics (e.g., {"new_cases_found": 10, "pages_processed": 5})
        error_message: Description of the error that caused the failure

    Returns:
        Success status (bool)
    """
    if not job_run_id:
        print("Warning: Cannot fail job, no job_run_id provided.")
        return False

    try:
        update_data = {
            "end_time": time.strftime("%Y-%m-%d %H:%M:%S%z"),
            "status": "failed",
            "error_message": error_message,
            **metrics,
        }

        response = (
            supabase_client.table("cron_job_runs")
            .update(update_data)
            .eq("id", job_run_id)
            .execute()
        )

        return bool(response.data)
    except Exception as e:
        print(f"Error marking cron job as failed: {e}")
        return False
